### PR TITLE
nicovideo.jp support

### LINKED ISSUE #60: NicoNico support
This will be useful because other downloaders are hard to find and it might just not work.

### Some research

NicoNico video pages has authorization data for guest user in HTML attribute with ``id="js-initial-watch-data"``.

Front-end web sends to POST request to ``https://api.dmc.nico/api/sessions?_format=json`` (The DMC service; Modern delivery system) with JSON body of some data from authorization data.

The DMC service will respond with authorized URL to the master M3U8. And yes, NicoNico uses HLS VOD...

### Preview URLs
- https://www.nicovideo.jp/watch/sm13876659
  - Can be without WWW in start of the hostname
- https://sp.nicovideo.jp/watch/sm6287843
  - Poster mode

### PR DESCRIPTION
**Pull request by synzr**
### progress
- [x] manage to replicate the domand requests
- [x] implement it with no account requirement
- [x] ~~implement it with account requirement (not necessary)~~

closes #60

### CODE REVIEW

**Comment by synzr** [comment]
waiting for double internal stream filter fix

image
image

**Review by lostdusty — COMMENTED** [review]
LGTM

**Review by yeeterlol — COMMENTED** [review]
LGTM, broken on my machine when downloading but I think it should work fine if I don't have skill issues. All tests do pass.

File: src/modules/processing/services/nicovideo.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    (variant) => variant.resolution.height === height
+  );
+
+  if (hlsContent === undefined) {
+    hlsContent = hls.variants
+      .sort(
+        (firstVariant, secondVariant) =>
+          firstVariant.bandwidth - secondVariant.bandwidth
+      )
+      .shift();
```
Comment (dumbmoron) [inline src/modules/processing/services/nicovideo.js:133]: could this be .reduce instead or am i missing something (above too)

Comment (synzr) [reply src/modules/processing/services/nicovideo.js:133]: yep, reduce() fits much better

```javascript
// getting the HQ audio
const { id: audioId } = audios
  .filter((audio) => audio.isAvailable)
  .reduce(
    (firstAudio, secondAudio) =>
      firstAudio.bitRate > secondAudio.bitRate
        ? firstAudio
        : secondAudio
  )
```

**Comment by synzr** [comment]
> LGTM, broken on my machine when downloading. All tests do pass and I confirmed with @synzr that it works on his machine so I think it's just minor skill issues.

the ffmpeg-static just hangs with no error log when cobalt passes delivery.domand.nicovideo.jp URLs.
try to install ffmpeg manually and provide its path via FFMPEG_BIN environment variable.

**Comment by synzr** [comment]
> > LGTM, broken on my machine when downloading. All tests do pass and I confirmed with @synzr that it works on his machine so I think it's just minor skill issues.
> 
> the ffmpeg-static just hangs with no error log when cobalt passes delivery.domand.nicovideo.jp URLs. try to install ffmpeg manually and provide its absolute path via FFMPEG_BIN environment variable.

btw i can implement internal streams for m3u8 content to fix this

**Comment by yeeterlol** [comment]
> > LGTM, broken on my machine when downloading. All tests do pass and I confirmed with @synzr that it works on his machine so I think it's just minor skill issues.
> 
> the ffmpeg-static just hangs with no error log when cobalt passes delivery.domand.nicovideo.jp URLs. try to install ffmpeg manually and provide its absolute path via FFMPEG_BIN environment variable.

nvm the issue was that i forgot to install nscd since you need to install it on debian 12 too?

**Comment by synzr** [comment]
> > > LGTM, broken on my machine when downloading. All tests do pass and I confirmed with @synzr that it works on his machine so I think it's just minor skill issues.
> > 
> > 
> > the ffmpeg-static just hangs with no error log when cobalt passes delivery.domand.nicovideo.jp URLs. try to install ffmpeg manually and provide its absolute path via FFMPEG_BIN environment variable.
> 
> nvm the issue was that i forgot to install nscd since you need to install it on debian 12 and the running an instance guide said ubuntu 22.04 would have this issue. probably a weird edge case idk

OH, IT'S WHY? i'm just using lmde rn.

**Comment by dumbmoron** [comment]
> > > LGTM, broken on my machine when downloading. All tests do pass and I confirmed with @synzr that it works on his machine so I think it's just minor skill issues.
> > 
> > 
> > the ffmpeg-static just hangs with no error log when cobalt passes delivery.domand.nicovideo.jp URLs. try to install ffmpeg manually and provide its absolute path via FFMPEG_BIN environment variable.
> 
> btw i can implement internal streams for m3u8 content to fix this

if this fixes it can u do that but in a separate PR

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Comment by yeeterlol** [comment]
Nicovideo.jp is down for a while due to a cyberattack 
https://www.theregister.com/2024/06/11/niconico_ebiten_kadokawa_cyberattack_outages/

**Comment by RyutoSetsujin** [comment]
Nicovideo.jp is currently up in a very basic form, only allowing you to watch the most popular videos from 2007 and 2008. You must either live in Japan or have a Japanese VPN to access the site.
They are removing the videos from 2007 tomorrow.
https://x.com/nico_nico_info/status/1802546484753428867

**Comment by wukko** [comment]
old pr, can't merge or even rebase without rewriting the entire thing
